metric_learn/sdml.py

```diff
@@ -52,24 +55,50 @@ def __init__(self, balance_param=0.5, sparsity_param=0.01, use_cov=True,
     super(_BaseSDML, self).__init__(preprocessor)
 
   def _fit(self, pairs, y):
+    if not has_installed_skggm():
+      msg = ("Warning, skggm is not installed, so SDML will use "
+             "scikit-learn's graphical_lasso method. It can fail to converge"
+             "on some non SPD matrices where skggm would converge. If so, "
+             "try to install skggm. (see the README.md for the right "
+             "version.)")
+      warnings.warn(msg)
+    else:
+      print("SDML will use skggm's solver.")
     pairs, y = self._prepare_inputs(pairs, y,
                                     type_of_inputs='tuples')
 
     # set up prior M
     if self.use_cov:
       X = np.vstack({tuple(row) for row in pairs.reshape(-1, pairs.shape[2])})
-      M = pinvh(np.atleast_2d(np.cov(X, rowvar = False)))
+      prior = pinvh(np.atleast_2d(np.cov(X, rowvar=False)))
     else:
-      M = np.identity(pairs.shape[2])
+      prior = np.identity(pairs.shape[2])
     diff = pairs[:, 0] - pairs[:, 1]
     loss_matrix = (diff.T * y).dot(diff)
-    P = M + self.balance_param * loss_matrix
-    emp_cov = pinvh(P)
-    # hack: ensure positive semidefinite
-    emp_cov = emp_cov.T.dot(emp_cov)
-    _, M = graph_lasso(emp_cov, self.sparsity_param, verbose=self.verbose)
-
-    self.transformer_ = transformer_from_metric(M)
+    emp_cov = pinvh(prior) + self.balance_param * loss_matrix
+
+    # our initialization will be the matrix with emp_cov's eigenvalues,
+    # with a constant added so that they are all positive (plus an epsilon
+    # to ensure definiteness). This is empirical.
+    w, V = np.linalg.eigh(emp_cov)
+    if any(w < 0.):
+      warnings.warn("Warning, the input matrix of graphical lasso is not "
+                    "positive semi-definite (PSD). The algorithm may diverge, "
+                    "and lead to degenerate solutions. "
+                    "To prevent that, try to decrease the balance parameter "
+                    "`balance_param` and/or to set use_covariance=False.",
+                    ConvergenceWarning)
+    sigma0 = (V * (w - min(0, np.min(w)) + 1e-10)).dot(V.T)
```

Comment:
Maybe simpler:

```python
min_eigval = w.min()
if min_eigval < 0:
  warnings.warn(...)
  min_eigval = 0

w += 1e-10 - min_eigval
sigma0 = (V * w).dot(V.T)
```

Reply:
In the heuristic if `min_eigval > 0` we would want to keep the matrix untouched (with just an epsilon added), (so I don't think with your solution it would be the case since we would substract the min eigval) but I agree something simpler would be better, so something like:

```python
min_eigval = w.min()
if min_eigval < 0:
  warnings.warn(...)
  w -= min_eigval   # we translate the eigenvalues to make them all positive

w += 1e-10  # we add a small offset to avoid definiteness problems
sigma0 = (V * w).dot(V.T)
```

Reply:
done